### DISCUSSION TITLE
Auth, extend, docs [v5]

### DISCUSSION THREAD

**Discussion by sasha172017** [adonisjs/core#2649] [Help]
DOCS: "guards.driver
The driver to use for login and authenticating users. You can use one of the pre-existing drivers or extend the Auth module to add your own."

How can write extend? and where write? please,details,thank

**Comment by guntribam** [comment]
I think this is the way:

```ts
import { ApplicationContract } from '@ioc:Adonis/Core/Application'

export default class AppProvider {
  constructor(protected app: ApplicationContract) {}

  public register() {
  }

  public async boot() {
    // IoC container is ready
    const auth = this.app.container.use('Adonis/Addons/Auth')
    auth.extend(type, provider, callback) //right here!!

  }

  public async ready() {
    // App is ready
  }

  public async shutdown() {
    // Cleanup, since app is going down
  }
}

```

**Reply by guntribam** [reply]
I'm facing a similiar issue cause i need to search the username + password in LDAP before get the user data from the database. My solution was to extend the auth module, but that was taking a long time, so i just perform the ldap authentication before my BASIC auth guard authentication inside the middleware.

**Reply by sasha172017** [reply]
what is auth.extend(type, provider, callback) ? what does return callback? do you have example this method? thank

**Reply by guntribam** [reply]
no i don't
